contracts/InnerMajoritySet.sol

```diff
@@ -0,0 +1,299 @@
+// Once given, support can be removed.
+// Addresses supported by more than half of the existing validators are the validators.
+// Malicious behaviour causes support removal.
+// Benign misbehaviour causes supprt removal if its called again after MAX_INACTIVITY.
+// Benign misbehaviour can be absolved before being called the second time.
+
+contract InnerMajoritySet is InnerSet {
+	// EVENTS
+	event Report(address indexed reporter, address indexed reported, bool indexed malicious);
+	event Support(address indexed supporter, address indexed supported, bool indexed added);
+	event InitiateChange(bytes32 indexed _parent_hash, address[] _new_set);
+	event ChangeFinalized(address[] current_set);
+
+	struct ValidatorStatus {
+		// Is this a validator.
+		bool isValidator;
+		// Index in the validatorList.
+		uint index;
+		// Validator addresses which supported the address.
+		AddressVotes.Data support;
+		// Keeps track of the votes given out while the address is a validator.
+		address[] supported;
+		// Initial benign misbehaviour time tracker.
+		mapping(address => uint) firstBenign;
+		// Repeated benign misbehaviour counter.
+		AddressVotes.Data benignMisbehaviour;
+	}
+
+	// Support can not be added once this number of validators is reached.
+	uint public constant MAX_VALIDATORS = 30;
+	// Time after which the validators will report a validator as malicious.
+	uint public constant MAX_INACTIVITY = 6 hours;
+	// Ignore misbehaviour older than this number of blocks.
+	uint public constant RECENT_BLOCKS = 20;
+
+	// STATE
+
+	// Current list of addresses entitled to participate in the consensus.
+	address[] public validatorsList;
+	// Pending list of validator addresses.
+	address[] pendingList = [0x00f5777f8133aae2734396ab1d43ca54ad11bfb737];
+	// Tracker of status for each address.
+	mapping(address => ValidatorStatus) validatorsStatus;
+
+	// Used to lower the constructor cost.
+	AddressVotes.Data initialSupport;
+
+	// Each validator is initially supported by all others.
+	constructor() public {
+		initialSupport.count = pendingList.length;
+		for (uint i = 0; i < pendingList.length; i++) {
+			address supporter = pendingList[i];
+			initialSupport.inserted[supporter] = true;
+		}
+
+		for (uint j = 0; j < pendingList.length; j++) {
+			address validator = pendingList[j];
+			validatorsStatus[validator] = ValidatorStatus({
+				isValidator: true,
+				index: j,
+				support: initialSupport,
+				supported: pendingList,
+				benignMisbehaviour: AddressVotes.Data({ count: 0 })
+			});
+		}
+		validatorsList = pendingList;
+	}
+
+	// Called on every block to update node validator list.
+	function getValidators() public constant returns (address[]) {
+		return validatorsList;
+	}
+
+	// Log desire to change the current list.
+	function initiateChange() private {
+		outerSet.initiateChange(blockhash(block.number - 1), pendingList);
+		emit InitiateChange(blockhash(block.number - 1), pendingList);
```

Comment:
the linters will report all sorts of errors for the `emit`

see https://github.com/protofire/solhint/issues/33